### PR TITLE
Init allArchives set in ApplicationArchivesBuildItem constructor

### LINKED ISSUE #43138: Race in ApplicationArchivesBuildItem.java getAllApplicationArchives()
### Describe the bug

I am an undergraduate student at the University of Crete and as part of my BSc thesis, I have created a tool that dynamically detects data races in Java. The tool is similar to LLVM's TSan and provides informative traces of the origin of each race. The tool is not currently open source but I am planning to make it, soon.

I have used my tool to run some of the integration tests of Quarkus and have detected a race in `ApplicationArchivesBuildItem.java` line, **49**, while running the `openapi` test.

### Expected behavior

No race reported

### Actual behavior

Stack trace of the race

```
==================
WARNING: ThreadSanitizer: data race (pid=3958727)
Write of size 8 at 0x7f3b5d58d420 by thread T63:
# true race
  #0 io.quarkus.deployment.builditem.ApplicationArchivesBuildItem.getAllApplicationArchives() ApplicationArchivesBuildItem.java:49
  #1 io.quarkus.deployment.steps.CombinedIndexBuildStep.build() CombinedIndexBuildStep.java:33
  #2 jdk.internal.reflect.NativeMethodAccessorImpl.invoke0() NativeMethodAccessorImpl.java:-1
  #3 jdk.internal.reflect.NativeMethodAccessorImpl.invoke() NativeMethodAccessorImpl.java:56
  #4 jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke() DelegatingMethodAccessorImpl.java:43
  #5 java.lang.reflect.Method.invoke() Method.java:558
  #6 io.quarkus.deployment.ExtensionLoader$3.execute() ExtensionLoader.java:792
  #7 io.quarkus.builder.BuildContext.run() BuildContext.java:243
  #8 io.quarkus.builder.BuildContext$$Lambda$1870/0x00007f345699da90.run() <null>:-1
  #9 org.jboss.threads.ContextHandler$1.runWith() ContextHandler.java:18
  #10 org.jboss.threads.EnhancedQueueExecutor$Task.run() EnhancedQueueExecutor.java:2504
  #11 org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run() EnhancedQueueExecutor.java:1507
  #12 java.lang.Thread.run() Thread.java:843
  #13 org.jboss.threads.JBossThread.run() JBossThread.java:499
 Previous read of size 8 at 0x7f3b5d58d420 by thread T66:
  #0 io.quarkus.deployment.builditem.ApplicationArchivesBuildItem.getAllApplicationArchives() ApplicationArchivesBuildItem.java:46
  #1 io.quarkus.arc.deployment.SplitPackageProcessor.splitPackageDetection() SplitPackageProcessor.java:58
  #2 jdk.internal.reflect.NativeMethodAccessorImpl.invoke0() NativeMethodAccessorImpl.java:-1
  #3 jdk.internal.reflect.NativeMethodAccessorImpl.invoke() NativeMethodAccessorImpl.java:56
  #4 jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke() DelegatingMethodAccessorImpl.java:43
  #5 java.lang.reflect.Method.invoke() Method.java:558
  #6 io.quarkus.deployment.ExtensionLoader$3.execute() ExtensionLoader.java:792
  #7 io.quarkus.builder.BuildContext.run() BuildContext.java:243
  #8 io.quarkus.builder.BuildContext$$Lambda$1870/0x00007f345699da90.run() <null>:-1
  #9 org.jboss.threads.ContextHandler$1.runWith() ContextHandler.java:18
  #10 org.jboss.threads.EnhancedQueueExecutor$Task.run() EnhancedQueueExecutor.java:2504

SUMMARY: ThreadSanitizer: data race ApplicationArchivesBuildItem.java:49 in io.quarkus.deployment.builditem.ApplicationArchivesBuildItem.getAllApplicationArchives()
==================
```

### How to Reproduce?

MAVEN_OPTS="-XX:+MaTSa" ./mvnw verify -pl integration-tests/openapi -Dstart-containers -Dtest-containers 

### Output of `uname -a` or `ver`

Linux pc8 5.4.0-193-generic #213-Ubuntu SMP Fri Aug 2 19:14:16 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux

### Output of `java -version`

openjdk version "17.0.11-internal" 2024-04-16 OpenJDK Runtime Environment (build 17.0.11-internal+0-adhoc.aantonak.jdk17u0117) OpenJDK 64-Bit Server VM (build 17.0.11-internal+0-adhoc.aantonak.jdk17u0117, mixed mode)

### Quarkus version or git rev

3.8

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

apache-maven-3.9.3

### PR DESCRIPTION
**Pull request by alx01011**
Closes #43138

### CODE REVIEW

File: core/deployment/src/main/java/io/quarkus/deployment/builditem/ApplicationArchivesBuildItem.java [hunk 1]
```diff
@@ -20,6 +20,11 @@ public final class ApplicationArchivesBuildItem extends SimpleBuildItem {
     public ApplicationArchivesBuildItem(ApplicationArchive root, Collection<ApplicationArchive> applicationArchives) {
         this.root = root;
         this.applicationArchives = applicationArchives;
+
+        HashSet<ApplicationArchive> ret = new HashSet<>(applicationArchives);
+        ret.add(root);
+
+        this.allArchives = Collections.unmodifiableSet(ret);
```
Comment (mkouba) [inline core/deployment/src/main/java/io/quarkus/deployment/builditem/ApplicationArchivesBuildItem.java:27]: You still need to make the `allArchives` field final.

Comment (alx01011) [reply core/deployment/src/main/java/io/quarkus/deployment/builditem/ApplicationArchivesBuildItem.java:27]: You are right. Done.

**Review by mkouba — COMMENTED** [review]
Looks good but pls squash the commits ;-)

**Comment by mkouba** [comment]
> Looks good but pls squash the commits ;-)

@alx01011 I meant _squash_ as in rebasing in a single commit...

**Comment by alx01011** [comment]
> > Looks good but pls squash the commits ;-)
> 
> @alx01011 I meant _squash_ as in rebasing in a single commit...

Yes I got that, I am on it. Sorry for the mess...

**Comment by mkouba** [comment]
> > > Looks good but pls squash the commits ;-)
> > 
> > 
> > @alx01011 I meant _squash_ as in rebasing in a single commit...
> 
> Yes I got that, I am on it. Sorry for the mess...

No problem at all!

**Review by zakkak — COMMENTED** [review]
I would expect this PR to only touch https://github.com/quarkusio/quarkus/pull/43143/files#diff-2b34fb8257e0e2796c68ce0ec7046ecea0a2f643e3a4c1314c51b74f297f2fe8

Are the changes in the rest of the files intentional?

@alx01011 @mkouba

**Comment by mkouba** [comment]
> I would expect this PR to only touch https://github.com/quarkusio/quarkus/pull/43143/files#diff-2b34fb8257e0e2796c68ce0ec7046ecea0a2f643e3a4c1314c51b74f297f2fe8
> 
> Are the changes in the rest of the files intentional?
> 
> @alx01011 @mkouba

This PR is closed and was superseded by https://github.com/quarkusio/quarkus/pull/43145.

**Comment by zakkak** [comment]
Oh, sorry for the noise. Followed the wrong link...

**Pull request closed without merging**